src/mlpack/methods/cf/normalization/combined_normalization.hpp

```diff
@@ -0,0 +1,208 @@
+  template<
+      int I, /* Which normalization in tuple to use */
+      typename MatType,
+      typename = std::enable_if_t<(I < std::tuple_size<TupleType>::value)>>
+  void SequenceNormalize(MatType& data)
+  {
+    std::get<I>(normalizations).Normalize(data);
+    SequenceNormalize<I+1>(data);
+  }
+
+  //! End of tuple unpacking.
+  template<
+      int I, /* Which normalization in tuple to use */
+      typename MatType,
+      typename = std::enable_if_t<(I >= std::tuple_size<TupleType>::value)>,
+      typename = void>
+  void SequenceNormalize(MatType& /* data */) { }
+
+  //! Unpack normalizations tuple to denormalize.
+  template<
+      int I, /* Which normalization in tuple to use */
+      typename = std::enable_if_t<(I < std::tuple_size<TupleType>::value)>>
+  double SequenceDenormalize(const size_t user,
+                             const size_t item,
+                             const double rating) const
+  {
+    // The order of denormalization should be the reversed order
+    // of normalization.
+    double realRating = SequenceDenormalize<I+1>(user, item, rating);
+    realRating =
+        std::get<I>(normalizations).Denormalize(user, item, realRating);
+    return realRating;
+  }
+
+  //! End of tuple unpacking.
+  template<
+      int I, /* Which normalization in tuple to use */
+      typename = std::enable_if_t<(I >= std::tuple_size<TupleType>::value)>,
+      typename = void>
+  double SequenceDenormalize(const size_t /* user */,
+                             const size_t /* item */,
+                             const double rating) const
+  {
+    return rating;
+  }
+
+  //! Unpack normalizations tuple to denormalize.
+  template<
+      int I, /* Which normalization in tuple to use */
+      typename = std::enable_if_t<(I < std::tuple_size<TupleType>::value)>>
+  void SequenceDenormalize(const arma::Mat<size_t>& combinations,
+                           arma::vec& predictions) const
+  {
+    // The order of denormalization should be the reversed order
+    // of normalization.
+    SequenceDenormalize<I+1>(combinations, predictions);
+    std::get<I>(normalizations).Denormalize(combinations, predictions);
+  }
+
+  //! End of tuple unpacking.
+  template<
+      int I, /* Which normalization in tuple to use */
+      typename = std::enable_if_t<(I >= std::tuple_size<TupleType>::value)>,
+      typename = void>
+  void SequenceDenormalize(const arma::Mat<size_t>& /* combinations */,
+                           arma::vec& /* predictions */) const { }
+
+  //! Unpack normalizations tuple to serialize.
+  template<
+      int I, /* Which normalization in tuple to serialize */
+      typename Archive,
+      typename = std::enable_if_t<(I < std::tuple_size<TupleType>::value)>>
+  void SequenceSerialize(Archive& ar, const unsigned int version)
+  {
+    std::string tagName = "normalization_";
+    tagName += std::to_string(I);
+    ar & boost::serialization::make_nvp(
```

Comment:
Correct me if I'm mistaken, but why don't you write just `ar & BOOST_SERIALIZATION_NVP(normalizations);`? I thought we can serialize tuples.

Reply:
Boost Serialization doesn't support tuple serialization (refer to [here](https://stackoverflow.com/questions/14744303/does-boost-support-serialization-of-c11s-stdtuple)), so I wrote an implementation to serialize it.

Reply:
Yeah, you're right. Curious, Boost 1.66 still doesn't support tuples, I didn't know.

Reply:
Yea boost 1.66 still doesn't support tuple serialization. I tried and it fails.